runtime/compiler/wasm/reader_test.go

```diff
@@ -0,0 +1,825 @@
+
+	t.Parallel()
+
+	read := func(data []byte) ([]*FunctionType, error) {
+		b := buf{data: data}
+		r := WASMReader{buf: &b}
+		err := r.readTypeSection()
+		if err != nil {
+			return nil, err
+		}
+		return r.Module.Types, nil
+	}
+
+	t.Run("valid", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 7 (LEB128)
+			0x87, 0x80, 0x80, 0x80, 0x0,
+			// type count
+			0x1,
+			// function type
+			0x60,
+			// parameter count: 2
+			0x2,
+			// type of parameter 1: i32
+			0x7f,
+			// type of parameter 2: i32
+			0x7f,
+			// return value count: 1
+			0x1,
+			// type of return value 1: i32
+			0x7f,
+		})
+		require.NoError(t, err)
+		assert.Equal(t,
+			[]*FunctionType{
+				{
+					Params:  []ValueType{ValueTypeI32, ValueTypeI32},
+					Results: []ValueType{ValueTypeI32},
+				},
+			},
+			funcTypes,
+		)
+	})
+
+	t.Run("invalid size", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			0x87, 0x80, 0x80,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidSectionSizeError{
+				Offset:    0,
+				ReadError: io.EOF,
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid count", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 7 (LEB128)
+			0x80, 0x80, 0x80, 0x80, 0x0,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidTypeSectionTypeCountError{
+				Offset:    5,
+				ReadError: io.EOF,
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid indicator", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 2 (LEB128)
+			0x82, 0x80, 0x80, 0x80, 0x0,
+			// type count
+			0x1,
+			// function type
+			0xFF,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidFuncTypeIndicatorError{
+				Offset:            6,
+				FuncTypeIndicator: 0xff,
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid parameter count", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 4 (LEB128)
+			0x82, 0x80, 0x80, 0x80, 0x0,
+			// type count
+			0x1,
+			// function type
+			0x60,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidFuncTypeParameterCountError{
+				Offset:    7,
+				ReadError: io.EOF,
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid parameter type", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 4 (LEB128)
+			0x84, 0x80, 0x80, 0x80, 0x0,
+			// type count
+			0x1,
+			// function type
+			0x60,
+			// parameter count: 1
+			0x1,
+			// type of parameter 1
+			0xff,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidFuncTypeParameterTypeError{
+				Index: 0,
+				ReadError: InvalidValTypeError{
+					Offset:  8,
+					ValType: 0xFF,
+				},
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid, parameter type missing", func(t *testing.T) {
```

Comment:
👍 

This is a case when not enough parameter type is provided, I wonder if too many parameter types are provided, what error would it return? I guess the extra parameter type would be parsed as the next item, which is "return value count", which would shows a different error message?